README.md
# Red Hat Typeface Files

## Designers

### Jeremy Mickel

Jeremy Mickel runs [MCKL](https://www.mckltype.com), a Los Angeles-based type foundry and design studio publishing original fonts and creating custom designs for clients. Founded in 2012, MCKL has collaborated with leading design firms, companies, and organizations around the world to provide custom typeface and logo design services. Mickel's work has been recognized by the Type Directors Club and the AIGA, and he has taught at RISD and the Minneapolis College of Art and Design.

## About Red Hat Display and Red Hat Text

![Type specimen](type-specimen@2x.png)

Red Hat is an enterprise software company with an open source development model. We use collaboration and knowledge sharing to craft better, more reliable, and more adaptable technologies. How our words look is as important to our brand voice as the words we choose. That’s why we developed a type family that’s all our own.

The Red Hat Typeface is a superfamily of Display, Text, and Mono styles, each with a range of weights in roman and italic. The fonts were originally commissioned by Paula Scher / [Pentagram](https://www.pentagram.com/) and designed by Jeremy Mickel / [MCKL](https://www.mckltype.com) for the new Red Hat identity.

Red Hat is a fresh take on the geometric sans genre, taking inspiration from a range of American sans serifs including Tempo and Highway Gothic. The Display styles, made for headlines and big statements, are low contrast and spaced tightly, with a large x-height and open counters. The Text styles have a slightly smaller x-height and narrower width for better legibility, are spaced more generously, and have thinned joins for better performance at small sizes. In 2021 we added Light and Light Italic styles, and a Monospace family. The fonts can be used together seamlessly at a range of sizes.

As part of Red Hat’s commitment to open source software, the fonts are made available for use under the SIL Open Font License.

## Variable Fonts

Variable fonts are available for each of the Red Hat Typeface families. The fonts include the `wght` axis, which allows for interpolation between light and black weights.

A demo for variable fonts is available at [https://redhatofficial.github.io/RedHatFont/](https://redhatofficial.github.io/RedHatFont/).

## Building the Fonts

From terminal, run the build script at `sources/build-all.sh`. Fonts output to `fonts/`.

NOTE: The first time you build, you will need to set up a virtual environment and install dependencies:

<details>
<summary><b><!-------->Setting up the build environment<!--------></b> (Click to expand)</summary>

### Set up the environment

**The basics**

You will need to open a terminal to run the following commands.

Clone the repo & navigate into it:

```
git clone https://github.com/RedHatOfficial/RedHatFont.git
cd RedHatFont
```

Check that you have Python 3:

```
which python3
```

It should return a path ending with `python3`, such as `/Library/Frameworks/Python.framework/Versions/3.7/bin/python3`. If it returns an error like `python3 not found`, you will need to [download Python 3](https://www.python.org/downloads/).

**Setting up a virtual environment**

To build, set up the virtual environment:

```bash
cd ~
python3 -m venv venv
```

Then activate it:

```bash
source venv/bin/activate
```

Now, install requirements:

```bash
cd RedHatFont
pip install -U -r requirements.txt
```

Give the build scripts permission to run/execute (you can copy & paste, then run both lines in the terminal at once):

```bash
chmod +x build-scripts/*.sh
```

**Making woff2 files**

Finally, you will also need to separately install [google/woff2](https://github.com/google/woff2) to enable the `woff2_compress` and `woff2_decompress` commands. Open a new terminal session, window, or tab to do this step.

```bash
# open a new terminal session first, then run
git clone --recursive https://github.com/google/woff2.git
cd woff2
make clean all
```

To make sure woff2_compress is installed properly, enter the following inyour terminal window:

```
woff2_compress
```

If terminal cannot find the command, you may need to ensure binaries are in $PATH, [a description of which you can find here.](https://github.com/google/woff2/issues/131)

Once woff2_compress is working in your terminal, you can now run the build!

</details>

### Build fonts

Once you have set up the environment (see above), you can build fonts & prep releases!

To build variable and static fonts, plus make woff2s, use `build-all.sh`. This takes awhile (most of the time is taken up by building TTF & OTF static fonts).

```bash
./build-all.sh
```

If you just want to build variable fonts, use `build.sh`:

```bash
./build-scripts/build-vf.sh
```

To build only the static fonts (these are secondary to the variable fonts, so you can’t set the version numbers in this script), use `build-statics.sh`:

```bash
./build-scripts/build-statics.sh
```

## Installation

The OTF or TTF folders contain the font files used by most user operating systems.

If you are running Fedora, Red Hat Enterprise Linux 7, CentOS 7, or any similar derivatives, you can install the fonts with the following:
```
sudo yum install redhat-display-fonts redhat-text-fonts
```
Note that Red Hat Enterprise Linux/CentOS users will need to [enable Fedora EPEL first](https://fedoraproject.org/wiki/EPEL).


If you are running Homebrew, you can install the fonts with the following:

```text
brew cask install homebrew/cask-fonts/font-redhat
```

## Bug reports and improvement requests

If you find a problem with a font file or have a request for future development of a font project, please [create a new issue in this project's issue tracker](https://github.com/RedHatOfficial/RedHatFont/issues).

## Self-Host Fonts Available From Red Hat

Since all the fonts available here are licensed with permission to redistribute, subject to the license terms, you are able to self-host the fonts in this project.

## Licensing

Copyright 2021 Red Hat, Inc.

Licensed under the SIL Open Font License, Version 1.1, with Reserved Font Name Red Hat.

The SIL OFL does not grant any rights under trademark law and all such rights are reserved. Modified versions must be renamed to avoid use of any Red Hat trademarks, including but not limited to "Red Hat".


CHANGELOG.md
## 4.0.2 (2021-04-28)
Tag: [4.0.2](https://github.com/RedHatOfficial/RedHatFont/releases/tag/4.0.2)

- Fixed red-hat-font.css to bring weights back to what they were in v3.
- Updated demo page to only use variable fonts
- Added demo page of modified variable fonts

## 4.0.1 (2021-04-22)
Tag: [4.0.1](https://github.com/RedHatOfficial/RedHatFont/releases/tag/4.0.1)

- Refreshed metainfo files to account for the new "Red Hat Typeface" name and monospace font family

## 4.0.0 (2021-04-08)
Tag: [4.0.0](https://github.com/RedHatOfficial/RedHatFont/commits/4.0.0)

- New weights (Light / Light Italic) in Red Hat Display and Red Hat Text
- Tabular figures (monospaced numbers) in Red Hat Display and Red Hat Text
- Red Hat Mono typeface
- Variable font versions of all the fonts
- Other small improvements to glyph shape, kerning, metadata across the entire family of fonts
- Added new fonts to webfont css file
- Created new variable webfont css file
- Added minified versions of webfont css files
- Updated demo page with variable fonts
- Dropped support for IE11.

## 3.0.0 (2021-01-15)
Tag: [3.0.0](https://github.com/RedHatOfficial/RedHatFont/commits/3.0.0)

- Update CSS to reference woff2 files
- Remove the unused `text-rendering` property from font-face
- Add `font-display` property to font-face for better performance on slow connections
- Drop support for older browsers (IE8)

## 2.3.2 (2019-07-28)
Tag: [2.3.2](https://github.com/RedHatOfficial/RedHatFont/commits/2.3.2)

- Add font preview image to README
- Remove `_DS_Store` files
- Add information about font installation
- Add basic AppStream metainfo data

## 2.3.1 (2019-07-16)
Tag: [2.3.1](https://github.com/RedHatOfficial/RedHatFont/commits/2.3.1)

- Set the OS/2 fsSelection bit 6 to on for both the TTF and OTF fonts that are not Roman, Bold, Italic or Bold Italic
- Turned off setting the OS/2 fsSelection bit 9 for italic fonts (the Oblique bit)
- Bumped font version to 1.005

## 2.3.0 (2019-06-24)
Tag: [2.3.0](https://github.com/RedHatOfficial/RedHatFont/commits/2.3.0)

- Revised the italic ‘fontinfo’ files, to meet Google Font standards
- Added the build.py, build_ttf.py, ttf_postprocess.py files for building the mastered fonts

## 2.2.0 (2019-04-25)
Tag: [2.2.0](https://github.com/RedHatOfficial/RedHatFont/commits/2.2.0)

- README update

## 2.1.0 (2019-04-25)
Tag: [2.1.0](https://github.com/RedHatOfficial/RedHatFont/commits/2.1.0)

- Add source files to repo to to support Google Fonts

## 2.0.0 (2019-04-17)
Tag: [2.0.0](https://github.com/RedHatOfficial/RedHatFont/commits/2.0.0)

- Update font family variations, name change to italic versions for Google fonts
- Update sass mixin: remove SVG reference, update relative path default, remove quotations from font-style
- Update license

## 1.0.0 (2019-03-27)
Tag: [1.0.0](https://github.com/RedHatOfficial/RedHatFont/commits/1.0.0)

- The "Red Hat Display" and "Red Hat Text" font families which include regular, medium, & bold weights, as well as the italic style.


source_files/build.py
import sys
import os 
import os.path
import tempfile
import re
from fontTools.ttLib import TTFont
from fontTools import ttLib
from fontTools.ttLib.tables._k_e_r_n import KernTable_format_0


def makeotf(outlineSourcePath, featuresPath=None, glyphOrderPath=None, menuNamePath=None, fontInfoPath=None, releaseMode=True, setOS2Bit6=False):
    """
    Run makeotf.
    The arguments will be converted into arguments
    for makeotf as follows:

    =================  ===
    outlineSourcePath  -f
    featuresPath       -ff
    glyphOrderPath     -gf
    menuNamePath       -mf
    fontInfoPath       -fi
    releaseMode        -r
    =================  ===
    """
    cmds = ["makeotf", "-nshw", "-sp", "-f", outlineSourcePath]
    if featuresPath:
        cmds.extend(["-ff", featuresPath])
    if glyphOrderPath:
        cmds.extend(["-gf", glyphOrderPath])
    if menuNamePath:
        cmds.extend(["-mf", menuNamePath])
    if fontInfoPath:
        cmds.extend(["-fi", fontInfoPath])
    if releaseMode:
        cmds.append("-r")
    if setOS2Bit6:
        cmds.extend(["-osbOn", "6"])
        cmds.extend(["-osbOn", "8"])
    stderr, stdout = _execute(cmds)
    return stderr, stdout

def autohint(fontPath):
    """
    Run autohint.
    The following arguments will be passed to autohint.

    * -nb
    * -a
    * -r
    * -q
    """
    cmds = ["autohint", "-nb", "-a", "-r", "-q", fontPath]
    stderr, stdout = _execute(cmds)
    return stderr, stdout

def lookUpDirTree(fileName):
    """ 
    This is called when we are using a default name for either the FontMenuNamDB or the GlyphOrderAndAliasDB files.
    These are often located one or two dir levels above the font file, as they are shared by the font family.
    """

    maxLevels = 2
    i = 0
    found = 0
    dirPath, fileName = os.path.split(fileName)
    while i <= maxLevels:
        path = os.path.join(dirPath,fileName)
        if os.path.exists(path):
            found = 1
            break
        dirPath = os.path.join(dirPath, "..")
        i += 1
    if not found:
        path = None
    return path

def getNameChanges(path):
    p = os.path.join(path, 'GlyphOrderAndAliasDB')
    goaadb = lookUpDirTree(p)
    if goaadb is not None:
        d = {}
        with open(goaadb) as f:
            for line in f:
               if len(line) > 1:
                   val = line.split()[0]
                   key = line.split()[1]
                   d[key] = val
        return d
    else:
        return {}

def subsetFlatKerning(kerning, path):
    plinc = ['space', 'exclam', 'quotesingle', 'quotedbl', 'numbersign', 'dollar', 'percent', 'ampersand', 'parenleft', 'parenright', 'asterisk', 'plus', 'comma', 'hyphen', 'period', 'slash', 'zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'colon', 'semicolon', 'less', 'equal', 'greater', 'question', 'at', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'bracketleft', 'backslash', 'bracketright', 'asciicircum', 'underscore', 'grave', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'braceleft', 'bar', 'braceright', 'asciitilde', 'exclamdown', 'cent', 'sterling', 'currency', 'yen', 'brokenbar', 'section', 'dieresis', 'copyright', 'ordfeminine', 'guillemotleft', 'logicalnot', 'registered', 'macron', 'degree', 'plusminus', 'twosuperior', 'threesuperior', 'acute', 'mu', 'paragraph', 'periodcentered', 'cedilla', 'onesuperior', 'ordmasculine', 'guillemotright', 'onequarter', 'onehalf', 'threequarters', 'questiondown', 'Agrave', 'Aacute', 'Acircumflex', 'Atilde', 'Adieresis', 'Aring', 'AE', 'Ccedilla', 'Egrave', 'Eacute', 'Ecircumflex', 'Edieresis', 'Igrave', 'Iacute', 'Icircumflex', 'Idieresis', 'Eth', 'Ntilde', 'Ograve', 'Oacute', 'Ocircumflex', 'Otilde', 'Odieresis', 'multiply', 'Oslash', 'Ugrave', 'Uacute', 'Ucircumflex', 'Udieresis', 'Yacute', 'Thorn', 'germandbls', 'agrave', 'aacute', 'acircumflex', 'atilde', 'adieresis', 'aring', 'ae', 'ccedilla', 'egrave', 'eacute', 'ecircumflex', 'edieresis', 'igrave', 'iacute', 'icircumflex', 'idieresis', 'eth', 'ntilde', 'ograve', 'oacute', 'ocircumflex', 'otilde', 'odieresis', 'divide', 'oslash', 'ugrave', 'uacute', 'ucircumflex', 'udieresis', 'yacute', 'thorn', 'ydieresis', 'dotlessi', 'circumflex', 'caron', 'breve', 'dotaccent', 'ring', 'ogonek', 'tilde', 'hungarumlaut', 'quoteleft', 'quoteright', 'minus']
    
    nameChanges = getNameChanges(path)
    
    new_plinc = []
    for x in plinc:
        if x in nameChanges.keys():
            new_plinc.append(nameChanges[x])
        else:
            new_plinc.append(x)
    new_kerning = {}
    
    for key, value in kerning.items():
        old = len(value)
        new_values = {}
        for p, v in value.items():
            if p[0] in new_plinc and p[1] in new_plinc:
                new_values[p] = v
        print 'Subset kerning'
        print 'Old pairs were: ' + str(old) +'. New pairs are: ' + str(len(new_values))
        new_kerning[key] = new_values
    return new_kerning


def versionClean(font):
    from fontTools.ttLib.tables._n_a_m_e import NameRecord, table__n_a_m_e
    
    nameIDs = [(5, 1, 0, 0), (5, 3, 1, 1033)]
    nameTable = font["name"]
    
    for n in nameIDs:
        nameRecord = nameTable.getName(n[0], n[1], n[2], n[3])
        s = nameRecord.string
        nameTable.names.remove(nameRecord)
        
        parts = s.split(';')
        
        s = ''
        for p in parts:
            if p != 'PS (version unavailable)':
                s = s + p + ';'
        s = s[:-1]
        
        record = NameRecord()
        record.nameID = n[0]
        record.platformID = n[1]
        record.platEncID = n[2]
        record.langID = n[3]
        record.string = s
        nameTable.names.append(record)

def nameTableTweak(font):
    nameIDs = [(16, 1, 0, 0), (17, 1, 0, 0)]
    nameTable = font["name"]
    
    for n in nameIDs:
        nameRecord = nameTable.getName(n[0], n[1], n[2], n[3])
        if nameRecord is not None:
            nameTable.names.remove(nameRecord)

def makeDSIG(font):
    from fontTools.ttLib.tables.D_S_I_G_ import SignatureRecord
    newDSIG = ttLib.newTable("DSIG")
    newDSIG.ulVersion = 1
    newDSIG.usFlag = 1
    newDSIG.usNumSigs = 1
    sig = SignatureRecord()
    sig.ulLength = 20
    sig.cbSignature = 12
    sig.usReserved2 = 0
    sig.usReserved1 = 0
    sig.pkcs7 = '\xd3M4\xd3M5\xd3M4\xd3M4'
    sig.ulFormat = 1
    sig.ulOffset = 20
    newDSIG.signatureRecords = [sig]
    font.tables["DSIG"] = newDSIG        

def _flat_kern_count(glyph, keys):
    count = 0
    for x in keys:
        if x[0] == glyph:
            count += 1
    return count

def processFont(path, d, flatKernData):
    font = TTFont(path)
    oldD, f = os.path.split(path)
    new = os.path.join(d, f)
    
    name = font['name'].getName(6, 1, 0, 0).string
    
    # Make a flat kerning table
    if flatKernData is not None:
        newKern = ttLib.newTable("kern")
        newKern.version = 0
        newKern.kernTables = [KernTable_format_0()]
    
    
        flat = flatKernData[name]
        sortedKeys = flat.keys()
        sortedKeys.sort()
    
    
        left = ''
        i = 0
        for key in sortedKeys:
            if left == '':
                left = key[0]
                kern_table = {key:flat[key]}
                count = _flat_kern_count(left, sortedKeys)
                i += 1
            elif i+1 < len(sortedKeys) and left != sortedKeys[i+1][0]:
                if count + _flat_kern_count(sortedKeys[i+1][0], sortedKeys) > 10920:
                    kern_table[key] = flat[key]
                
                    t = len(newKern.kernTables) - 1
                
                    table = newKern.kernTables[t]
                    table.kernTable = kern_table
                    table.version = 0
                    table.coverage = 1
                    table.apple = False
                    newKern.kernTables.append(KernTable_format_0())
                    left = ''
                else:                    
                    kern_table[key] = flat[key]
                    left = sortedKeys[i+1][0]
                    count = _flat_kern_count(sortedKeys[i+1][0], sortedKeys) + count
                i += 1
            elif len(sortedKeys) == i + 1:
                kern_table[key] = flat[key]
                t = len(newKern.kernTables) - 1
                table = newKern.kernTables[t]
                table.kernTable = kern_table
                table.version = 0
                table.coverage = 1
                table.apple = False
                i += 1
            else:
                kern_table[key] = flat[key]
                i += 1

        # Add this table back to the font
        font.tables["kern"] = newKern
    
    # Fix the ugly id string
    versionClean(font)
    nameTableTweak(font)
    makeDSIG(font)
    

    font.save(new)


def getFiles(path, extension):
    if not extension.startswith('.'):
        extension = '.' + extension
    if extension == '.ufo':
        return [dir for (dir, dirs, files) in os.walk(path) if dir[-len(extension):] == extension]
    else:
        return [os.sep.join((dir, file)) for (dir, dirs, files) in os.walk(path) for file in files if file[-len(extension):] == extension]

# --------------
# Internal Tools
# --------------

if sys.platform == "darwin":
    _fdkToolDirectory = os.path.join(os.environ["HOME"], "bin/FDK/Tools/osx")
else:
    _fdkToolDirectory = None

def _makeEnviron():
    env = dict(os.environ)
    if _fdkToolDirectory not in env["PATH"].split(":"):
        env["PATH"] += (":%s" % _fdkToolDirectory)
    kill = ["ARGVZERO", "EXECUTABLEPATH", "PYTHONHOME", "PYTHONPATH", "RESOURCEPATH"]
    for key in kill:
        if key in env:
            del env[key]
    return env

def _execute(cmds):
    import subprocess
    # for some reason, autohint and/or checkoutlines
    # locks up when subprocess.PIPE is given. subprocess
    # requires a real file so StringIO is not acceptable
    # here. thus, make a temporary file.
    stderrPath = tempfile.mkstemp()[1]
    stdoutPath = tempfile.mkstemp()[1]
    stderrFile = open(stderrPath, "w")
    stdoutFile = open(stdoutPath, "w")
    # get the os.environ
    env = _makeEnviron()
    # make a string of escaped commands
    cmds = subprocess.list2cmdline(cmds)
    # go
    popen = subprocess.Popen(cmds, stderr=stderrFile, stdout=stdoutFile, env=env, shell=True)
    popen.wait()
    # get the output
    stderrFile.close()
    stdoutFile.close()
    stderrFile = open(stderrPath, "r")
    stdoutFile = open(stdoutPath, "r")
    stderr = stderrFile.read()
    stdout = stdoutFile.read()
    stderrFile.close()
    stdoutFile.close()
    # trash the temp files
    os.remove(stderrPath)
    os.remove(stdoutPath)
    # done
    return stderr, stdout


    
def main():
    # make output dir
    d = os.getcwd() + '/fonts'
    try:
        os.makedirs(d)
    except OSError:
        if not os.path.isdir(d):
            raise
    
    # make sure output dir contains no otfs
    files = getFiles(d, 'otf')
    if len(files) != 0:
        for file in files:
            os.remove(file)
    
    print "Make Initial OTFs"
    print '-----------------------'
    
    # Make initial OTFs
    print os.getcwd()
    files = getFiles(os.getcwd(), 'ufo')
    print files
    result = ""
    for file in files:
        print file
        six = False
        font_style = file.split("-")[-1][:-4]
        if font_style not in ["Regular", "Italic", "Bold", "BoldItalic"]:
            if "Italic" not in font_style:
                six = True
        if six:
            stderr, stdout = makeotf(file, setOS2Bit6=True)
        else:
            stderr, stdout = makeotf(file)
        result += '------------------\nOutput for: ' + file + '\n\n'+ stderr + stdout
    path = os.path.join(os.getcwd(), 'makeOTF_report.txt')
    print path
    file = open(path, 'w')
    file.write(result)
    file.close()
    
    # TTX Magic
    print "TTX Magic Time"
    print '-----------------------'
    files = getFiles(os.getcwd(), 'otf')
    for file in files:
        print file
        try:
            current_directory = os.path.dirname(file)
            kerning_file = os.path.join(current_directory, 'flatKerning.py')
            with open(kerning_file,'r') as inf:
                flat_kerning = eval(inf.read())
            print 'Got flat kerning'
        except:
            flat_kerning = None
            print 'No flat kerning found'
        subset = os.path.join(current_directory, 'subset')
        if os.path.exists(subset):
            flat_kerning = subsetFlatKerning(flat_kerning, current_directory)
            
        processFont(file, d, flat_kerning)
    
    # Remove otfs used in the TTX operation
    for f in files:
        os.remove(f)
    
    
    # Autohint the ttxed otfs
    print "Autohint time"
    print '-----------------------'
    files = getFiles(d, 'otf')
    result = ""
    for file in files:
        print file
        stderr, stdout = autohint(file)
        result += '------------------\nOutput for: ' + file + '\n\n'+ stderr + stdout
    path = os.path.join(os.getcwd(), 'autohint_report.txt')
    print path
    file = open(path, 'w')
    file.write(result)
    file.close()    
    
if __name__ == "__main__":
    main()
